File: backend/core/services/exchange_rate_service.py
import requests
from django.core.cache import cache
from django.conf import settings
from datetime import datetime
from typing import Dict, Optional

class ExchangeRateService:
    API_URL = "https://api.exchangerate-api.com/v4/latest/"
    CACHE_TIMEOUT = 86400  # 24 horas en segundos
    CACHE_KEY_PREFIX = "exchange_rates_"
    REQUEST_TIMEOUT = 10  # Timeout en segundos para requests HTTP

    @classmethod
    def get_exchange_rates(cls, base_currency='USD') -> Dict:
        """
        Obtiene las tasas de cambio para una moneda base.
        Primero intenta obtenerlas del caché, si no existen, consulta la API externa.
        
        Args:
            base_currency: Código de moneda base (default: 'USD')
            
        Returns:
            Dict con las tasas de cambio o error estructurado
        """
        cache_key = f"{cls.CACHE_KEY_PREFIX}{base_currency}"
        cached_data = cache.get(cache_key)

        if cached_data:
            return cached_data

        try:
            response = requests.get(
                f"{cls.API_URL}{base_currency}",
                timeout=cls.REQUEST_TIMEOUT
            )
            response.raise_for_status()
            data = response.json()
            
            # Agregar timestamp de última actualización
            data['last_updated'] = datetime.utcnow().isoformat()
            data['base_currency'] = base_currency
            
            # Guardar en caché
            cache.set(cache_key, data, cls.CACHE_TIMEOUT)
            return data
            
        except requests.Timeout:
            # Intentar retornar datos antiguos del caché si existen
            stale_data = cache.get(cache_key)
            if stale_data:
                stale_data['stale'] = True
                stale_data['error'] = 'Timeout fetching fresh rates, using cached data'
                return stale_data
            return {
                "error": "Request timeout",
                "details": f"Could not fetch exchange rates within {cls.REQUEST_TIMEOUT} seconds",
                "base_currency": base_currency
            }
            
        except requests.RequestException as e:
            # Intentar retornar datos antiguos del caché si existen
            stale_data = cache.get(cache_key)
            if stale_data:
                stale_data['stale'] = True
                stale_data['error'] = 'Error fetching fresh rates, using cached data'
                return stale_data
            return {
                "error": "Could not fetch exchange rates",
                "details": str(e),
                "base_currency": base_currency
            }

    @classmethod
    def convert_price(cls, amount: float, from_currency: str, to_currency: str) -> Dict:
        """
        Convierte un precio de una moneda a otra.
        
        Args:
            amount: Cantidad a convertir
            from_currency: Moneda origen (ej: 'COP')
            to_currency: Moneda destino (ej: 'USD', 'EUR')
            
        Returns:
            Dict con el precio convertido o error
        """
        if amount < 0:
            return {
                "error": "Invalid amount",
                "details": "Amount must be positive"
            }
        
        if from_currency == to_currency:
            return {
                "original_amount": amount,
                "converted_amount": amount,
                "from_currency": from_currency,
                "to_currency": to_currency,
                "rate": 1.0
            }
        
        # Obtener tasas de cambio (usando USD como base común)
        rates_data = cls.get_exchange_rates('USD')
        
        if "error" in rates_data:
            return rates_data
        
        rates = rates_data.get('rates', {})
        
        # Si la moneda origen es USD, usar tasa directa
        if from_currency == 'USD':
            if to_currency not in rates:
                return {
                    "error": "Currency not supported",
                    "details": f"Currency {to_currency} not found in exchange rates"
                }
            rate = rates[to_currency]
            converted = amount * rate
        # Si la moneda destino es USD, usar inversa
        elif to_currency == 'USD':
            if from_currency not in rates:
                return {
                    "error": "Currency not supported",
                    "details": f"Currency {from_currency} not found in exchange rates"
                }
            rate = 1 / rates[from_currency]
            converted = amount * rate
        # Conversión entre dos monedas no-USD
        else:
            if from_currency not in rates or to_currency not in rates:
                return {
                    "error": "Currency not supported",
                    "details": f"One or both currencies ({from_currency}, {to_currency}) not found in exchange rates"
                }
            # Convertir: origen -> USD -> destino
            from_rate = rates[from_currency]
            to_rate = rates[to_currency]
            rate = to_rate / from_rate
            converted = amount * rate
        
        return {
            "original_amount": amount,
            "converted_amount": round(converted, 2),
            "from_currency": from_currency,
            "to_currency": to_currency,
            "rate": round(rate, 6),
            "last_updated": rates_data.get('last_updated')
        }

File: frontend/src/services/exchangeRateService.js
import api from '../api/axiosGlobalInstance';

/**
 * Servicio para obtener tasas de cambio y convertir precios
 */
export const getExchangeRates = (baseCurrency = 'USD') =>
  api.get('/exchange-rates/', { params: { base: baseCurrency } })
    .then((res) => res.data)
    .catch((error) => {
      console.error('Error fetching exchange rates:', error);
      // Si es un error de red, retornar un objeto de error estructurado
      if (!error.response) {
        return {
          error: 'Network error',
          details: 'Could not connect to the server. Please check if the backend is running.'
        };
      }
      throw error;
    });

/**
 * Convierte un precio de una moneda a otra
 * @param {number} amount - Cantidad a convertir
 * @param {string} fromCurrency - Moneda origen (ej: 'COP')
 * @param {string} toCurrency - Moneda destino (ej: 'USD', 'EUR')
 * @returns {Promise} Promise con el resultado de la conversión
 */
export const convertPrice = (amount, fromCurrency, toCurrency) =>
  api.post('/exchange-rates/convert/', {
    amount,
    from_currency: fromCurrency,
    to_currency: toCurrency
  })
    .then((res) => {
      // El backend ya devuelve rate correctamente calculado
      // Usar directamente los valores del backend
      if (res.data && res.data.rate !== undefined) {
        return {
          rate: res.data.rate,
          converted_amount: res.data.converted_amount,
          original_amount: res.data.original_amount,
          from_currency: res.data.from_currency,
          to_currency: res.data.to_currency
        };
      }
      return res.data;
    })
    .catch((error) => {
      console.error('Error converting price:', error);
      // Si es un error de red o 401, retornar un objeto de error estructurado
      if (!error.response || error.response.status === 401) {
        return {
          error: error.response?.status === 401 
            ? 'Unauthorized: Please check backend permissions' 
            : 'Network error',
          details: 'Could not connect to the server. Please check if the backend is running.',
          converted_amount: null,
          rate: 1 // Fallback rate
        };
      }
      // Para otros errores, retornar objeto de error en lugar de lanzar
      return {
        error: error.response?.data?.error || 'Conversion failed',
        details: error.response?.data?.details || error.message,
        converted_amount: null,
        rate: 1 // Fallback rate
      };
    });

/**
 * Obtiene múltiples conversiones de precio en paralelo
 * @param {number} amount - Cantidad a convertir
 * @param {string} fromCurrency - Moneda origen
 * @param {string[]} toCurrencies - Array de monedas destino
 * @returns {Promise} Promise con objeto de conversiones { USD: {...}, EUR: {...}, ... }
 */
export const convertPriceToMultiple = async (amount, fromCurrency, toCurrencies) => {
  const conversions = await Promise.allSettled(
    toCurrencies.map(currency => convertPrice(amount, fromCurrency, currency))
  );
  
  const result = {};
  toCurrencies.forEach((currency, index) => {
    if (conversions[index].status === 'fulfilled') {
      result[currency] = conversions[index].value;
    } else {
      result[currency] = {
        error: conversions[index].reason?.response?.data?.error || 'Conversion failed',
        converted_amount: null
      };
    }
  });
  
  return result;
};

